### ISSUE TITLE
Interface for randomness

### ISSUE THREAD

**Issue by chebee7i** [networkx/networkx#1550]
In various places, we make use of random number generators. Due to our pure Python tilt, we've tended to use the `random` module. Some functions accept seeds as input, others (I have not checked) might not provide that option at all.

This briefly came up in https://github.com/networkx/networkx/issues/1341#issuecomment-76645971, and maybe it's time to dig in a bit further.

My opinion is that we should not use the `random` module. While it has satisfied our needs thus far, it does not allow enough flexibility since 1) there is always a global state and 2) there aren't many distributions provided. The ability to have different random number generator instances requires that we move away from using the `random` module.

Instead, I think any function that requires randomness should have a hard dependency on NumPy and that we should standardize on NumPy's `RandomState` API. This isn't that much of a burden since most installations can easily obtain NumPy, and a good deal of NetworkX now relies on NumPy anyway. This is also what _most_ Python packages do if they require randomness. See for example, [scikit-learn-#random-numbers](http://scikit-learn.org/stable/developers/#random-numbers). The idea is that functions/classes/methods will have a parameter, `prng` or `random_state` (as in `sklearn`) or `seed`.  We then use that to retrieve a random number generator using something identical to [check_random_state](http://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html#sklearn.utils.check_random_state) (but with a better name). Then the function uses that random number generator. This will make code more robust and more easily tested.
#1523 will be affected by this decision.

**Comment by hagberg** [comment]
It's a reasonable idea.  A couple thoughts -
- The dependence on Numpy is all  (or mostly) through "linear algebra" approaches to graph problems so this would add a general dependency in other places.
- Without making it overly complicated would it be possible to still use the built-in Python random generator?   That would be nice for e.g. pypy or other cases when Numpy is unavailable.

**Comment by ysitu** [comment]
Actually, you can instantiate the `random.Random` class. The way it is documented makes it look as if it were undocumented, but it is indeed documented.

**Comment by chebee7i** [comment]
Ah ok, then that should be doable!

**Comment by jfinkels** [comment]
When this change is made, a reminder should also be documented somewhere that randomness should not be used for cryptographic security (unless NetworkX explicitly includes a secure random number generator).

**Comment by NeilGirdhar** [comment]
I agree with most of the motivations and the initial comments, but please don't accept `seed` or `RandomState`.  Accept `rng` with a consistent interface.  This allows you to plug in any random number generator and sacrifices nothing.  In your code, you can default `rng` to being numpy's rng if you like.

**Comment by dschult** [comment]
Take a look at #1764 for more on this issue

**Comment by dschult** [comment]
See #1764 which will fix this using ideas fro #2766
